### PR TITLE
Add animations.

### PR DESCRIPTION
**Pull request by kattni**
Created `animations.py` and added animations and animation features:

- Animations: ColorCycle, Solid, Blink, Comet, Sparkle
- AnimationGroup - ability to add multiple LED instances (such as a strip and the CPB onboard LEDs) to a group to animate both simultaneously
- AnimationSequence - cycle through animations
- Animate features: fill, freeze, resume, change_color

Added RAINBOW color list to `color.py`.

### CODE REVIEW

File: adafruit_led_animation/animation.py [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+        for peer in self.peers:
+            peer.show()
+
+        self._next_update = now + self._speed_ns
+        return True
+
+    def draw(self):
+        """
+        Animation subclasses must implement draw() to render the animation sequence.
+        """
```
Comment (caternuson) [inline adafruit_led_animation/animation.py:112]: add a `raise NotImplementedError()` here instead of leaving it empty

File: adafruit_led_animation/animation.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+    # pylint: disable=too-many-arguments
+    def __init__(self, pixel_object, speed, color, peers=None, paused=False):
+        self.pixel_object = pixel_object
+        self._speed_ns = 0
+        self.speed = speed  # sets _speed_ns
+        self._color = color
+        self._next_update = monotonic_ns()
+        self.pixel_object.auto_write = False
+        self.color = color
+        self.peers = peers if peers else []
```
Comment (caternuson) [inline adafruit_led_animation/animation.py:75]: This could be done by just having the default for the parameter be `[]` instead of `None`

File: adafruit_led_animation/animation.py [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+    def speed(self, seconds):
+        self._speed_ns = int(seconds * 1000000000)
+
+    def _recompute_color(self, color):
+        pass
+
+    def change_color(self, color):
+        """
+        By default changing color uses the color property.
+        """
```
Comment (caternuson) [inline adafruit_led_animation/animation.py:174]: Why is this needed? Why not just use the `color` property setter?

File: adafruit_led_animation/animation.py [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+        index = 0
+        while True:
+            self._color = self.colors[index]
+            yield
+            index = (index + 1) % len(self.colors)
+
+    def change_color(self, color):
+        """
+        ColorCycle doesn't support change_color.
+        """
```
Comment (caternuson) [inline adafruit_led_animation/animation.py]: Should an exception be raised here?

File: adafruit_led_animation/animation.py [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+        self._paused = True
+        self._time_left_at_pause = max(0, monotonic_ns() - self._next_update)
+
+    def resume(self):
+        """
+        Resumes the animation.
+        """
+        self._next_update = monotonic_ns() + self._time_left_at_pause
+        self._time_left_at_pause = 0
+        self._paused = False
```
Comment (caternuson) [inline adafruit_led_animation/animation.py:134]: How about replacing `freeze()` and `resume()` with a `paused` property that can be set or return current state?
```python
# to pause
animation.paused = True
# to check
if animation.paused:
    print("Animation currently paused.")
# to un-pause
animation.paused = False
```

File: adafruit_led_animation/animation.py (hunk 2 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py:75]: Pylint hates [] as the default for a param.
`W: 70, 4: Dangerous default value [] as argument (dangerous-default-value)`

File: adafruit_led_animation/animation.py (hunk 1 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py:112]: Good catch.

File: adafruit_led_animation/animation.py (hunk 3 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py:174]: It's difficult to override a property setter.  I'm trying to remember why I needed it.  I think _recompute_color is what needs overriding and I forgot I put it there.  Will try to squash this.

Comment (rhooper) [reply adafruit_led_animation/animation.py:174]: Oh, now I remember - it ended up there because I didn't use a property on an animation sequence, time to fix that.

File: adafruit_led_animation/animation.py (hunk 4 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py]: Code removed.

File: adafruit_led_animation/animation.py (hunk 5 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py:134]: The freeze/resume protocol on the Animation children grew out of the AnimationSequence and AnimationGroup.  It would be possible to use a paused property, but it doesn't behave cleanly:

AnimationSequence freezes/resumes the current animation in the sequence.
AnimationGroup freezes/resumes all members of the group.

Querying AnimationSequence.paused would return the sequence paused state so that wouldn't be too bad.

Querying AnimationGroup.paused however would need to return all() or any() on the paused state of all members of the group - as it doesn't itself know if they're paused.

I intentionally made the basic signatures for Animations, AnimationGroup and AnimationSequence identical so you can mix and match them in any order.

Comment (caternuson) [reply adafruit_led_animation/animation.py:134]: So is the only potential confusion when querying the current state of an AnimationGroup? If freeze/resume acts on all members, how do you get to a state where they are mixed?

Comment (kattni) [reply adafruit_led_animation/animation.py:134]: Having a mixed state in an AnimationGroup seems to defeat the purpose of the Group. If your end goal is a mixed state, you'd want to handle things separately.

@caternuson What is your reasoning behind suggesting this change? I'm trying to understand the purpose of the suggestion.

Comment (rhooper) [reply adafruit_led_animation/animation.py:134]: i guess the more important thing here is what do we want the interface to be for pausing and resuming. a property or two methods? @caternuson @kattni

Comment (caternuson) [reply adafruit_led_animation/animation.py:134]: Was thinking it would be more Pythonic. Also, to have a way to get current state. But this could be refactored later.

**Review by caternuson — APPROVED** [review]

File: adafruit_led_animation/animation.py (hunk 5 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py:134]: Refactoring it later not the right choice as it would change the protocol people use.... So I'll get this refactored ASAP, as @kattni needs this for her latest guide.

Comment (caternuson) [reply adafruit_led_animation/animation.py:134]: I'd say keep as is then. The audioio API has separate functions for play/stop/pause:
https://circuitpython.readthedocs.io/en/latest/shared-bindings/audioio/AudioOut.html
so seems OK.

File: adafruit_led_animation/animation.py [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+            color = (color >> 16 & 0xff, color >> 8 & 0xff, color & 0xff)
+        self._color = color
+        self._recompute_color(color)
+
+    @property
+    def speed(self):
+        """
+        The animation speed in fractional seconds.
+        """
+        return self._speed_ns / 1000000000
```
Comment (theacodes) [inline adafruit_led_animation/animation.py]: Since this is technically a "magic" number and appears more than once, consider pulling this into a module-level constant, like https://github.com/theacodes/Winterbloom-Sol-software/blob/ce2b8b07f19b5f5c20cd32fa50e2ef7e7007c603/winterbloom_sol/adsr.py#L30

**Comment by rhooper** [comment]
@caternuson @kattni Let me know which you prefer.

File: adafruit_led_animation/animation.py [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+    # pylint: disable=too-many-arguments
+    def __init__(self, pixel_object, speed, color, tail_length=10, reverse=False, bounce=False):
+        self._tail_length = tail_length + 1
+        self._color_step = 0.9 / tail_length
+        self._color_offset = 0.1
+        self._comet_colors = None
+        self._reverse_comet_colors = None
+        self.reverse = reverse
+        self.bounce = bounce
+        super(Comet, self).__init__(pixel_object, speed, color)
```
Comment (theacodes) [inline adafruit_led_animation/animation.py:256]: nit: it's unusual to see the superclass `__init__` so far down. I've generally seen this done as soon as possible, like in the very first statement in the `__init__` method.

**Comment by caternuson** [comment]
@rhooper Now that I've looked at the audioio API, what do you think about mimicking that? Is your animation control similar enough to audio control?

File: adafruit_led_animation/animation.py [hunk 8]
```diff
@@ -0,0 +1,585 @@
...
+        for pixel in range(len(self.pixel_object)):
+            if self.pixel_object[pixel] == self._half_color:
+                self.pixel_object[pixel] = half_color
+            elif self.pixel_object[pixel] == self._dim_color:
+                self.pixel_object[pixel] = dim_color
+        self._half_color = half_color
+        self._dim_color = dim_color
+
+    def draw(self):
+        pixel = random.randint(0, (len(self.pixel_object) - 2))
```
Comment (theacodes) [inline adafruit_led_animation/animation.py:325]: what if the pixel object has less than 2 pixels?

**Comment by rhooper** [comment]
@caternuson I'm not sure, I'll have to look at the AudioOut API when I have time.

**Comment by caternuson** [comment]
@rhooper @kattni If this is time critical, just go with what you had originally, the `freeze()` and `resume()` functions.

File: adafruit_led_animation/animation.py [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+        self._cycle_position = 0
+        super(Pulse, self).__init__(pixel_object, speed, color)
+
+    def draw(self):
+        now = monotonic_ns()
+        time_since_last_draw = (now - self._last_update) / 1000000000
+        self._last_update = now
+        self._cycle_position = (self._cycle_position + time_since_last_draw) % self._period
+        intensity = self.min_intensity + (
+            sin(self._radians_per_second * self._cycle_position) * self._intensity_delta)
```
Comment (theacodes) [inline adafruit_led_animation/animation.py]: sin's range is -1.0 to 1.0, so this could end up setting intensity to a negative number.

For example:
```python
min_intensity = 0.0
max_intensity = 1.0
intensity_delta = max_intensity - min_intensity  # 1.0

# assume radians per second and cycle position end up passing 1.5pi to sin.
intensity = min_intensity + (sin(1.5 * pi) * intensity_delta)
intensity = 0.0 + (-1.0 * 1.0)
intensity = -1.0
```

The pixel colors would then be multiplied by `-1` which likely isn't what we want.

You can solve this by moving sin's range to be `0 - 1.0` like this:

```python
def normalized_sin(angle):
    return 0.5 + (sin(angle) * 0.5)
```

File: adafruit_led_animation/animation.py (hunk 6 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py]: I had thought about that, but I am also concerned about performance.  I vaguely remember there's a constant thing in  CircuitPython that might be faster than a bare module variable, but I need to benchmark.  Having worked a lot in the CircuitPython core with a JLink and gdb lately makes me think about these things :)

I'll look into the performance of both.

File: adafruit_led_animation/animation.py (hunk 7 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py:256]: I usually do it first, but in this case, when I did this, the color setup work needed doing before the rest of super is run.  I'll check if that's still true.  If not, a comment would be useful.

File: adafruit_led_animation/animation.py (hunk 8 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py:325]: Then it would crash.  Good call.  (note: this was based on code from elsewhere, and this animation doesn't support single pixels -- __init__ should check that).

File: adafruit_led_animation/animation.py (hunk 6 above)
Comment (theacodes) [reply adafruit_led_animation/animation.py]: From my experience moving things to module-level variables doesn't impact performance, and might even help, considering that CP doesn't need to do an allocation for the module-level variable every time you use it.

If performance is a big concern we should do another pass at this once it's merged. There's lots of use of `for x in y`, `range`, and slicing that outright murder performance on CP compared to manipulating index variables.

File: adafruit_led_animation/animation.py (hunk 9 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py]: I'm planning to remove the use of sin() very soon, as the performance is poor.  The code, however, shouldn't produce a negative brightness because it's clamped to 0-180 degrees by self._radians_per_second and % self._period set during init.

File: adafruit_led_animation/animation.py (hunk 6 above)
Comment (kattni) [reply adafruit_led_animation/animation.py]: Another pass at it was already planned. The initial code was written to make it work. Making it better is the next step.

Comment (rhooper) [reply adafruit_led_animation/animation.py]: @theacodes there is indeed plenty of room for performance fixes.  Two different things happened while writing this:  1) Making it work.  2) A lot of the code I wrote was written with _pixelbuf in mind, which perform faster with slices..  Running the code without _pixelbuf is, as you said, much slower.  Once we get this merged, there's a number of issues to file regarding improvements, additions, and design documentation.

File: adafruit_led_animation/animation.py (hunk 5 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py:134]: Reverted back to the freeze/resume methods.

File: adafruit_led_animation/animation.py (hunk 6 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py]: Removed the use of sin, however the code is still too slow for large strips as-is, but not as slow as it was - but that'll be easy to address in future PRs.

File: adafruit_led_animation/animation.py (hunk 7 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py:256]: Comments added and init for both the base class and this class reorganized slightly.

File: adafruit_led_animation/animation.py (hunk 8 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py:325]: Init now checks that the strip is big enough.

File: adafruit_led_animation/animation.py (hunk 6 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py]: I moved this to a constant in the module.

File: adafruit_led_animation/animation.py (hunk 9 above)
Comment (rhooper) [reply adafruit_led_animation/animation.py]: Sin removed in favour of a linear function for now.  Code still needs significant optimization as  expected.

**Comment by kattni** [comment]
@theacodes Thank you for the review notes. @rhooper addressed them. As getting this merged is time-sensitive, I'm going to go ahead and merge it. If you have further suggestions, please file an issue. We intend to file separate issues for each of the improvements we want to make to make it easier to discuss them in the thread - please consider doing the same if you decide to file any issues. Thanks!
